File: tree树/avltree.py
#定义一个AVL树，实现了AVL树的自平衡功能，即在插入和删除节点时，自动调整树的结构，使其保持平衡

class AVLTreeNode(object):
    def __init__(self, value):
        self.value = value
        self.height = 0
        self.left = None
        self.right = None


class AVLTree(object):
    def __init__(self):
        self._root = None

    def _insert(self, node, value):
        if node is None:
            node = AVLTreeNode(value)
        elif value < node.value:
            node.left = self._insert(node.left, value)
            if (self.height(node.left) - self.height(node.right)) == 2:
                if value < node.left.value:
                    node = self._left_left_rotation(node)
                else:
                    node = self._left_right_rotation(node)
        elif value > node.value:
            node.right = self._insert(node.right, value)
            if (self.height(node.right) - self.height(node.left)) == 2:
                if value > node.right.value:
                    node = self._right_right_rotation(node)
                else:
                    node = self._right_left_rotation(node)
        node.height = max(self.height(node.left), self.height(node.right)) + 1
        return node

    def insert(self, value):
        self._root = self._insert(self._root, value)

    def _delete(self, node, value):
        if node is None:
            raise ValueError('value not found in the tree')
        elif value < node.value:
            node.left = self._delete(node.left, value)
            if (self.height(node.right) - self.height(node.left)) == 2:
                if node.right.left is not None and (self.height(node.right.left) > self.height(node.right.right)):
                    node = self._right_left_rotation(node)
                else:
                    node = self._right_right_rotation(node)
        elif value > node.value:
            node.right = self._delete(node.right, value)
            if (self.height(node.left) - self.height(node.right)) == 2:
                if node.left.right is not None and (self.height(node.left.right) > self.height(node.left.left)):
                    node = self._left_right_rotation(node)
                else:
                    node = self._left_left_rotation(node)
        elif node.left and node.right:
            if node.left.height <= node.right.height:
                min_node = self._find_min(node.right)
                node.value = min_node.value
                node.right = self._delete(node.right, min_node.value)
            else:
                max_node = self._find_max(node.left)
                node.value = max_node.value
                node.left = self._delete(node.left, max_node.value)
        else:
            if node.left:
                node = node.left
            else:
                node = node.right
        return node
    
    def delete(self, value):
        self._root = self._delete(self._root, value)

    def _find_min(self, node):
        if node.left is None:
            return node
        else:
            return self._find_min(node.left)
        
    def find_min(self):
        return self._find_min(self._root).value
    
    def _find_max(self, node):
        if node.right is None:
            return node
        else:
            return self._find_max(node.right)
        
    def find_max(self):
        return self._find_max(self._root).value
    
    def _left_left_rotation(self, node):
        k = node.left
        node.left = k.right
        k.right = node
        node.height = max(self.height(node.left), self.height(node.right)) + 1
        k.height = max(self.height(k.left), node.height) + 1
        return k
    
    def _right_right_rotation(self, node):
        k = node.right
        node.right = k.left
        k.left = node
        node.height = max(self.height(node.left), self.height(node.right)) + 1
        k.height = max(self.height(k.right), node.height) + 1
        return k
    
    def _left_right_rotation(self, node):
        node.left = self._right_right_rotation(node.left)
        return self._left_left_rotation(node)
    
    def _right_left_rotation(self, node):

        node.right = self._left_left_rotation(node.right)
        return self._right_right_rotation(node)
    
    def height(self, node):
        if node is None:
            return -1
        else:
            return node.height
        
    def _pre_order(self, node):
        if node is not None:
            print(node.value)
            self._pre_order(node.left)
            self._pre_order(node.right)

    def pre_order(self):
        self._pre_order(self._root)

    def _in_order(self, node):
        if node is not None:
            self._in_order(node.left)
            print(node.value)
            self._in_order(node.right)

    def in_order(self):
        self._in_order(self._root)

    def _post_order(self, node):
        if node is not None:
            self._post_order(node.left)
            self._post_order(node.right)
            print(node.value)

    def post_order(self):
        self._post_order(self._root)

    def _level_order(self, node):
        if node is None:
            return
        queue = []
        queue.append(node)
        while len(queue) > 0:
            node = queue.pop(0)
            print(node.value)
            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)

    def level_order(self):
        self._level_order(self._root)

    def _get(self, node, value):
        if node is None:
            return None
        if value < node.value:
            return self._get(node.left, value)
        elif value > node.value:
            return self._get(node.right, value)
        else:
            return node
        
    def get(self, value):
        return self._get(self._root, value)
    
    def _get_parent(self, node, value):
        if node is None:
            return None
        if value < node.value:
            if node.left.value == value:
                return node
            else:
                return self._get_parent(node.left, value)
        elif value > node.value:
            if node.right.value == value:
                return node
            else:
                return self._get_parent(node.right, value)
        else:
            return None
        
    def get_parent(self, value):
        return self._get_parent(self._root, value)
    
    def _get_min(self, node):
        if node.left is None:
            return node
        else:
            return self._get_min(node.left)
        
    def get_min(self):
        return self._get_min(self._root).value
    
    def _get_max(self, node):
        if node.right is None:
            return node
        else:
            return self._get_max(node.right)
        
    def get_max(self):
        return self._get_max(self._root).value
    

File: tree树/avltree.rs
//定义一个平衡二叉树，实现插入、删除、查找、遍历等操作

use std::cmp::Ordering;
use std::fmt::Debug;
use std::fmt::Display;
use std::fmt::Formatter;
use std::fmt::Result;
use std::mem::replace;
use std::ptr::NonNull;
use std::ptr::null_mut;

#[derive(Debug)]
pub struct AVLTree<T: Ord + Debug> {
    root: Option<NonNull<AVLNode<T>>>,
}

#[derive(Debug)]
struct AVLNode<T: Ord + Debug> {
    data: T,
    height: i32,
    left: Option<NonNull<AVLNode<T>>>,
    right: Option<NonNull<AVLNode<T>>>,
}

impl<T: Ord + Debug> AVLTree<T> {
    pub fn new() -> Self {
        AVLTree { root: None }
    }

    pub fn insert(&mut self, data: T) {
        let mut node = AVLNode::new(data);
        let mut parent = None;
        let mut cur = self.root;
        while let Some(mut cur_node) = cur {
            let cur_ref = unsafe { cur_node.as_mut() };
            parent = cur;
            if node.data < cur_ref.data {
                cur = cur_ref.left;
            } else {
                cur = cur_ref.right;
            }
        }
        if let Some(mut parent_node) = parent {
            let parent_ref = unsafe { parent_node.as_mut() };
            if node.data < parent_ref.data {
                parent_ref.left = Some(NonNull::new(&mut node as *mut _));
            } else {
                parent_ref.right = Some(NonNull::new(&mut node as *mut _));
            }
        } else {
            self.root = Some(NonNull::new(&mut node as *mut _));
        }
        self.rebalance();
    }

    pub fn remove(&mut self, data: T) {
        let mut parent = None;
        let mut cur = self.root;
        while let Some(mut cur_node) = cur {
            let cur_ref = unsafe { cur_node.as_mut() };
            if data < cur_ref.data {
                parent = cur;
                cur = cur_ref.left;
            } else if data > cur_ref.data {
                parent = cur;
                cur = cur_ref.right;
            } else {
                if cur_ref.left.is_some() && cur_ref.right.is_some() {
                    let mut min_parent = cur;
                    let mut min = cur_ref.right;
                    while let Some(mut min_node) = min {
                        let min_ref = unsafe { min_node.as_mut() };
                        if min_ref.left.is_some() {
                            min_parent = min;
                            min = min_ref.left;
                        } else {
                            cur_ref.data = replace(&mut min_ref.data, data);
                            cur = min;
                            parent = min_parent;
                            break;
                        }
                    }
                } else {
                    let mut child = cur_ref.left.or(cur_ref.right);
                    if let Some(mut parent_node) = parent {
                        let parent_ref = unsafe { parent_node.as_mut() };
                        if cur_ref.data < parent_ref.data {
                            parent_ref.left = child;
                        } else {
                            parent_ref.right = child;
                        }
                    } else {
                        self.root = child;
                    }
                    break;
                }
            }
        }
        self.rebalance();
    }

    pub fn find(&self, data: T) -> Option<&T> {
        let mut cur = self.root;
        while let Some(mut cur_node) = cur {
            let cur_ref = unsafe { cur_node.as_mut() };
            if data < cur_ref.data {
                cur = cur_ref.left;
            } else if data > cur_ref.data {
                cur = cur_ref.right;
            } else {
                return Some(&cur_ref.data);
            }
        }
        None
    }

    pub fn preorder_traverse<F>(&self, mut f: F)
    where
        F: FnMut(&T),
    {
        let mut stack = Vec::new();
        let mut cur = self.root;
        while cur.is_some() || !stack.is_empty() {
            while let Some(mut cur_node) = cur {
                let cur_ref = unsafe { cur_node.as_mut() };
                f(&cur_ref.data);
                stack.push(cur);
                cur = cur_ref.left;
            }
            if let Some(mut cur_node) = stack.pop() {
                let cur_ref = unsafe { cur_node.as_mut() };
                cur = cur_ref.right;
            }
        }
    }

    pub fn inorder_traverse<F>(&self, mut f: F)
    where
        F: FnMut(&T),
    {
        let mut stack = Vec::new();
        let mut cur = self.root;
        while cur.is_some() || !stack.is_empty() {
            while let Some(mut cur_node) = cur {
                let cur_ref = unsafe { cur_node.as_mut() };
                stack.push(cur);
                cur = cur_ref.left;
            }
            if let Some(mut cur_node) = stack.pop() {
                let cur_ref = unsafe { cur_node.as_mut() };
                f(&cur_ref.data);
                cur = cur_ref.right;
            }
        }
    }

    pub fn postorder_traverse<F>(&self, mut f: F)
    where
        F: FnMut(&T),
    {
        let mut stack = Vec::new();
        let mut cur = self.root;
        let mut last = None;
        while cur.is_some() || !stack.is_empty() {
            while let Some(mut cur_node) = cur {
                let cur_ref = unsafe { cur_node.as_mut() };
                stack.push(cur);
                cur = cur_ref.left;
            }
            if let Some(mut cur_node) = stack.last() {
                let cur_ref = unsafe { cur_node.as_mut() };
                if cur_ref.right.is_none() || cur_ref.right == last {
                    f(&cur_ref.data);
                    last = stack.pop();
                } else {
                    cur = cur_ref.right;
                }
            }
        }
    }

    fn rebalance(&mut self) {
        let mut stack = Vec::new();
        let mut cur = self.root;
        while cur.is_some() || !stack.is_empty() {
            while let Some(mut cur_node) = cur {
                let cur_ref = unsafe { cur_node.as_mut() };
                stack.push(cur);
                cur = cur_ref.left;
            }
            if let Some(mut cur_node) = stack.pop() {
                let cur_ref = unsafe { cur_node.as_mut() };
                let left_height = cur_ref.left.as_ref().map_or(-1, |node| unsafe {
                    let node_ref = node.as_ref();
                    node_ref.height
                });
                let right_height = cur_ref.right.as_ref().map_or(-1, |node| unsafe {
                    let node_ref = node.as_ref();
                    node_ref.height
                });
                cur_ref.height = left_height.max(right_height) + 1;
                if left_height - right_height > 1 {
                    let left = cur_ref.left.unwrap();
                    let left_ref = unsafe { left.as_ref() };
                    let left_left_height = left_ref.left.as_ref().map_or(-1, |node| unsafe {
                        let node_ref = node.as_ref();
                        node_ref.height
                    });
                    let left_right_height = left_ref.right.as_ref().map_or(-1, |node| unsafe {
                        let node_ref = node.as_ref();
                        node_ref.height
                    });
                    if left_left_height >= left_right_height {
                        self.rotate_right(&mut cur_node);
                    } else {
                        self.rotate_left(&mut left);
                        self.rotate_right(&mut cur_node);
                    }
                } else if right_height - left_height > 1 {
                    let right = cur_ref.right.unwrap();
                    let right_ref = unsafe { right.as_ref() };
                    let right_left_height = right_ref.left.as_ref().map_or(-1, |node| unsafe {
                        let node_ref = node.as_ref();
                        node_ref.height
                    });
                    let right_right_height = right_ref.right.as_ref().map_or(-1, |node| unsafe {
                        let node_ref = node.as_ref();
                        node_ref.height
                    });
                    if right_right_height >= right_left_height {
                        self.rotate_left(&mut cur_node);
                    } else {
                        self.rotate_right(&mut right);
                        self.rotate_left(&mut cur_node);
                    }
                }
                cur = cur_ref.right;
            }
        }
    }
    
    fn rotate_left(&mut self, node: &mut NonNull<AVLNode<T>>) {
        let mut node_ref = unsafe { node.as_mut() };
        let mut right = node_ref.right.unwrap();
        let mut right_ref = unsafe { right.as_mut() };
        node_ref.right = right_ref.left;
        right_ref.left = Some(NonNull::new(node as *mut _));
        node_ref.height = node_ref
            .left
            .as_ref()
            .map_or(-1, |node| unsafe { node.as_ref().height })
            .max(node_ref.right.as_ref().map_or(-1, |node| unsafe { node.as_ref().height }))
            + 1;
        right_ref.height = right_ref
            .left
            .as_ref()
            .map_or(-1, |node| unsafe { node.as_ref().height })
            .max(right_ref.right.as_ref().map_or(-1, |node| unsafe { node.as_ref().height }))
            + 1;
        if let Some(mut parent) = node_ref.right {
            let mut parent_ref = unsafe { parent.as_mut() };
            parent_ref.left = Some(NonNull::new(&mut node_ref as *mut _));
        } else {
            self.root = Some(NonNull::new(&mut node_ref as *mut _));
        }
        *node = right;
    }

    fn rotate_right(&mut self, node: &mut NonNull<AVLNode<T>>) {
        let mut node_ref = unsafe { node.as_mut() };
        let mut left = node_ref.left.unwrap();
        let mut left_ref = unsafe { left.as_mut() };
        node_ref.left = left_ref.right;
        left_ref.right = Some(NonNull::new(node as *mut _));
        node_ref.height = node_ref
            .left
            .as_ref()
            .map_or(-1, |node| unsafe { node.as_ref().height })
            .max(node_ref.right.as_ref().map_or(-1, |node| unsafe { node.as_ref().height }))
            + 1;
        left_ref.height = left_ref
            .left
            .as_ref()
            .map_or(-1, |node| unsafe { node.as_ref().height })
            .max(left_ref.right.as_ref().map_or(-1, |node| unsafe { node.as_ref().height }))
            + 1;
        if let Some(mut parent) = node_ref.left {
            let mut parent_ref = unsafe { parent.as_mut() };
            parent_ref.right = Some(NonNull::new(&mut node_ref as *mut _));
        } else {
            self.root = Some(NonNull::new(&mut node_ref as *mut _));
        }
        *node = left;
    }

    pub fn is_empty(&self) -> bool {
        self.root.is_none()
    }

    pub fn clear(&mut self) {
        self.root = None;
    }

    pub fn height(&self) -> i32 {
        self.root
            .as_ref()
            .map_or(-1, |node| unsafe { node.as_ref().height })
    }

    pub fn size(&self) -> usize {
        let mut size = 0;
        let mut stack = Vec::new();
        let mut cur = self.root;
        while cur.is_some() || !stack.is_empty() {
            while let Some(mut cur_node) = cur {
                let cur_ref = unsafe { cur_node.as_mut() };
                stack.push(cur);
                cur = cur_ref.left;
            }
            if let Some(mut cur_node) = stack.pop() {
                let cur_ref = unsafe { cur_node.as_mut() };
                size += 1;
                cur = cur_ref.right;
            }
        }
        size
    }

    pub fn is_balanced(&self) -> bool {
        let mut stack = Vec::new();
        let mut cur = self.root;
        while cur.is_some() || !stack.is_empty() {
            while let Some(mut cur_node) = cur {
                let cur_ref = unsafe { cur_node.as_mut() };
                stack.push(cur);
                cur = cur_ref.left;
            }
            if let Some(mut cur_node) = stack.pop() {
                let cur_ref = unsafe { cur_node.as_mut() };
                let left_height = cur_ref.left.as_ref().map_or(-1, |node| unsafe {
                    let node_ref = node.as_ref();
                    node_ref.height
                });
                let right_height = cur_ref.right.as_ref().map_or(-1, |node| unsafe {
                    let node_ref = node.as_ref();
                    node_ref.height
                });
                if (left_height - right_height).abs() > 1 {
                    return false;
                }
                cur = cur_ref.right;
            }
        }
        true
    }

    pub fn is_bst(&self) -> bool {
        let mut stack = Vec::new();
        let mut cur = self.root;
        let mut last = None;
        while cur.is_some() || !stack.is_empty() {
            while let Some(mut cur_node) = cur {
                let cur_ref = unsafe { cur_node.as_mut() };
                stack.push(cur);
                cur = cur_ref.left;
            }
            if let Some(mut cur_node) = stack.pop() {
                let cur_ref = unsafe { cur_node.as_mut() };
                if let Some(last_node) = last {
                    let last_ref = unsafe { last_node.as_ref() };
                    if cur_ref.data <= last_ref.data {
                        return false;
                    }
                }
                last = Some(cur_node);
                cur = cur_ref.right;
            }
        }
        true
    }

    pub fn is_avl(&self) -> bool {
        self.is_balanced() && self.is_bst()
    }

    pub fn to_vec(&self) -> Vec<&T> {
        let mut vec = Vec::new();
        self.inorder_traverse(|data| vec.push(data));
        vec
    }

    pub fn to_sorted_vec(&self) -> Vec<&T> {
        let mut vec = Vec::new();
        self.inorder_traverse(|data| vec.push(data));
        vec.sort();
        vec
    }

    pub fn to_string(&self) -> String
    where
        T: Display,
    {
        let mut string = String::new();
        self.inorder_traverse(|data| {
            string.push_str(&format!("{} ", data));
        });
        string
    }

    pub fn to_sorted_string(&self) -> String
    where
        T: Display,
    {
        let mut string = String::new();
        self.inorder_traverse(|data| {
            string.push_str(&format!("{} ", data));
        });
        string.pop();
        string
    }

    pub fn to_dot(&self) -> String
    where
        T: Display,
    {
        let mut string = String::new();
        string.push_str("digraph {\n");
        if let Some(mut root) = self.root {
            let root_ref = unsafe { root.as_mut() };
            string.push_str(&format!("  {};\n", root_ref.data));
            let mut stack = Vec::new();
            stack.push(root);
            while let Some(mut cur) = stack.pop() {
                let cur_ref = unsafe { cur.as_mut() };
                if let Some(left) = cur_ref.left {
                    let left_ref = unsafe { left.as_mut() };
                    string.push_str(&format!("  {} -> {};\n", cur_ref.data, left_ref.data));
                    stack.push(left);
                }
                if let Some(right) = cur_ref.right {
                    let right_ref = unsafe { right.as_mut() };
                    string.push_str(&format!("  {} -> {};\n", cur_ref.data, right_ref.data));
                    stack.push(right);
                }
            }
        }
        string.push_str("}");
        string
    }

    pub fn to_sorted_dot(&self) -> String
    where
        T: Display,
    {
        let mut string = String::new();
        string.push_str("digraph {\n");
        if let Some(mut root) = self.root {
            let root_ref = unsafe { root.as_mut() };
            string.push_str(&format!("  {};\n", root_ref.data));
            let mut stack = Vec::new();
            stack.push(root);
            while let Some(mut cur) = stack.pop() {
                let cur_ref = unsafe { cur.as_mut() };
                if let Some(left) = cur_ref.left {
                    let left_ref = unsafe { left.as_mut() };
                    string.push_str(&format!("  {} -> {};\n", cur_ref.data, left_ref.data));
                    stack.push(left);
                }
                if let Some(right) = cur_ref.right {
                    let right_ref = unsafe { right.as_mut() };
                    string.push_str(&format!("  {} -> {};\n", cur_ref.data, right_ref.data));
                    stack.push(right);
                }
            }
        }
        string.push_str("}");
        string
    }

    pub fn to_graphviz(&self, path: &str) -> std::io::Result<()> {
        use std::fs::File;
        use std::io::Write;
        let mut file = File::create(path)?;
        file.write_all(self.to_dot().as_bytes())?;
        Ok(())
    }

    pub fn to_sorted_graphviz(&self, path: &str) -> std::io::Result<()> {
        use std::fs::File;
        use std::io::Write;
        let mut file = File::create(path)?;
        file.write_all(self.to_sorted_dot().as_bytes())?;
        Ok(())
    }

    pub fn from_vec(vec: Vec<T>) -> Self {
        let mut tree = AVLTree::new();
        for data in vec {
            tree.insert(data);
        }
        tree
    }

    pub fn from_sorted_vec(vec: Vec<T>) -> Self {
        let mut tree = AVLTree::new();
        tree.root = AVLNode::from_sorted_vec(vec);
        tree
    }

    pub fn from_string(string: &str) -> Self
    where
        T: std::str::FromStr,
    {
        let mut tree = AVLTree::new();
        for data in string.split_whitespace() {
            tree.insert(data.parse().unwrap());
        }
        tree
    }

    pub fn from_sorted_string(string: &str) -> Self
    where
        T: std::str::FromStr,
    {
        let mut tree = AVLTree::new();
        tree.root = AVLNode::from_sorted_string(string);
        tree
    }

    pub fn from_graphviz(path: &str) -> std::io::Result<Self>
    where
        T: std::str::FromStr,
    {
        use std::fs::File;
        use std::io::Read;
        let mut file = File::open(path)?;
        let mut string = String::new();
        file.read_to_string(&mut string)?;
        Ok(AVLTree::from_dot(&string))
    }

    pub fn from_sorted_graphviz(path: &str) -> std::io::Result<Self>
    where
        T: std::str::FromStr,
    {
        use std::fs::File;
        use std::io::Read;
        let mut file = File::open(path)?;
        let mut string = String::new();
        file.read_to_string(&mut string)?;
        Ok(AVLTree::from_sorted_dot(&string))
    }

    pub fn from_dot(string: &str) -> Self
    where
        T: std::str::FromStr,
    {
        let mut tree = AVLTree::new();
        let mut lines = string.lines();
        lines.next();
        for line in lines {
            let mut iter = line.split_whitespace();
            let from = iter.next().unwrap();
            let to = iter.next().unwrap();
            tree.insert(from.parse().unwrap());
            tree.insert(to.parse().unwrap());
        }
        tree
    }

    pub fn from_sorted_dot(string: &str) -> Self
    where
        T: std::str::FromStr,
    {
        let mut tree = AVLTree::new();
        let mut lines = string.lines();
        lines.next();
        for line in lines {
            let mut iter = line.split_whitespace();
            let from = iter.next().unwrap();
            let to = iter.next().unwrap();
            tree.insert(from.parse().unwrap());
            tree.insert(to.parse().unwrap());
        }
        tree
    }

    pub fn from_preorder_vec(vec: Vec<T>) -> Self {
        let mut tree = AVLTree::new();
        tree.root = AVLNode::from_preorder_vec(vec);
        tree
    }

    pub fn from_preorder_string(string: &str) -> Self
    where
        T: std::str::FromStr,
    {
        let mut tree = AVLTree::new();
        for data in string.split_whitespace() {
            tree.insert(data.parse().unwrap());
        }
        tree
    }